File: algorithms/Quick Sort/C/code.c
#include <stdio.h>

int partition(int arr[], int low, int high)
{
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++)
    {
        if (arr[j] <= pivot)
        {
            i++;

            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;

    return i + 1;
}

void quick_sort(int arr[], int low, int high)
{
    if (low >= high)
    {
        return;
    }

    int pivot_index = partition(arr, low, high);

    quick_sort(arr, low, pivot_index - 1);
    quick_sort(arr, pivot_index + 1, high);
}

int main(void)
{
    int arr[] = {10, 7, 8, 9, 1, 5};
    int size = sizeof(arr) / sizeof(arr[0]);

    quick_sort(arr, 0, size - 1);

    for (int i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }

    printf("\n");

    return 0;
}

File: algorithms/Quick Sort/Python/code.py
def quick_sort(arr):
    def partition(low, high):
        pivot = arr[high]
        i = low - 1

        for j in range(low, high):
            if arr[j] <= pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]

        arr[i + 1], arr[high] = arr[high], arr[i + 1]

        return i + 1

    def sort(low, high):
        if low >= high:
            return

        pivot_index = partition(low, high)

        sort(low, pivot_index - 1)
        sort(pivot_index + 1, high)

    if len(arr) > 1:
        sort(0, len(arr) - 1)


arr = [10, 7, 8, 9, 1, 5]

quick_sort(arr)

print(arr)